### ISSUE TITLE
High-resolution updates `group` perf harms `degrees.rs` example.

### ISSUE THREAD

**Issue by frankmcsherry** [frankmcsherry/differential-dataflow#19]
The `degrees.rs` example computes and maintains the distribution of degrees in a continually changing random graph. It looks roughly like so:

```rust
let degrs = edges.map(|(src, _)| src)
                 .count();

let distr = degrs.map(|(_, cnt)| cnt)
    		 .count();
```

The current `group` implementation has quadratic behavior with respect to the number of distinct times per key, within each batch. With sufficiently large graphs, random changes are unlikely to touch many nodes within the same batch, but they are quite likely to touch the same *counts*, in the second `count` operator.

Here are some timings with the current code, for batch size one:

```
Echidnatron% cargo run --release --example degrees -- 10000 50000 1
worker 0, round 1 finished after Duration { secs: 0, nanos: 71173 }
worker 0, round 2 finished after Duration { secs: 0, nanos: 73827 }
worker 0, round 3 finished after Duration { secs: 0, nanos: 57993 }
...
```

We would hope that with batching the throughput would improve, and it does for a bit. Here are batch sizes ten,
```
Echidnatron% cargo run --release --example degrees -- 10000 50000 10
worker 0, round 10 finished after Duration { secs: 0, nanos: 336847 }
worker 0, round 20 finished after Duration { secs: 0, nanos: 243068 }
worker 0, round 30 finished after Duration { secs: 0, nanos: 238114 }
...
```
 one hundred,
```
Echidnatron% cargo run --release --example degrees -- 10000 50000 100
worker 0, round 100 finished after Duration { secs: 0, nanos: 1528624 }
worker 0, round 200 finished after Duration { secs: 0, nanos: 2137224 }
worker 0, round 300 finished after Duration { secs: 0, nanos: 2024237 }
...
```
and one thousand,
```
Echidnatron% cargo run --release --example degrees -- 10000 50000 1000
worker 0, round 1000 finished after Duration { secs: 0, nanos: 90445490 }
worker 0, round 2000 finished after Duration { secs: 0, nanos: 127967149 }
worker 0, round 3000 finished after Duration { secs: 0, nanos: 128467679 }
```

The average time to process each update in the first batch, for example, goes from 71us to 33us to 15us to 90us as we increase the batch size. That was looking pretty good, until we hit batch size one thousand. The quadratic horror is showing up here, and is probably also tainting the batch size one hundred execution as well.

This issue is non-trivial to fix, and involves engineering deep in `group`s bowels, but it should be done. This issue will stick around until there is a responsible fix.

**Comment by frankmcsherry** [comment]
For comparison, consider the `degrees.rs` example where we remove the second `count`. This should give us a sense for how things could be with an improved `group` implementation (a target). Note, this is doing less work, but the removed count is the one with a fairly compact set of keys, and *should* be the easier of the two.

I've just pasted all of the first batch times for increasing batch sizes together, using the same parameters as above:

```
worker 0, round 1 finished after Duration { secs: 0, nanos: 44841 }
worker 0, round 10 finished after Duration { secs: 0, nanos: 73881 }
worker 0, round 100 finished after Duration { secs: 0, nanos: 370281 }
worker 0, round 1000 finished after Duration { secs: 0, nanos: 1767591 }
```

The average times are 45us, 7.4us, 3.7us, and 1.8us. We would like the average times with the second `count` re-enabled to be less than twice these times, again because the second `count` is "less challenging" than the first for most other stream processors / for loops.

**Comment by frankmcsherry** [comment]
Further investigation suggests that while we can probably get the time to update collections down to something nearly linear, there is a bigger problem for general partial orders.

The current framework requires us to determine new "synthetic" times in the future when the collection might change due to interactions between the input updates, though not at the times of input updates. These are found by joining pairs of times, corresponding to the time at which two updates first *both* come into effect. 

Consider a batch with updates for a key at times `(0, i), (1, i-1), .. (i, 0)`. The set of pairwise joins of these times fill out half of the `(0,i)` x `(0,i)` square, and this is totally correct: the input collection can and does change at each of these times.

The issue here is that to discover these times we current "need to" join each update with each other update that is not less than it (which does allow us to optimize out times of accumulated updates, for example). This results in a quadratic amount of work that may or may not produce quadratically many times; if there are many times we don't feel bad (because it is the correct answer), but if not then we have spent a great deal of time that we "shouldn't have". 

For example, consider updates with times `(i, i % 2)`, which move forward through time and alternate iteration `0` and `1`. There are only a linear number of synthetic interesting times, but we risk doing quadratic work to discover this. The way to see it, that might guide an implementation, is that for each iteration the updates are totally ordered, and if we processed each as a totally ordered batch we should be better positioned to do it efficiently (no promises at this point).

**Comment by frankmcsherry** [comment]
One easy remedial fix is to specialize the implementations for total orders, which are something we can see and exploit at the type level. This would fix `degrees.rs` and a few other implementations, but leave uses `group` inside `iterate` somewhat mysterious.

**Comment by frankmcsherry** [comment]
I have a candidate `group` implementation which exploits totally ordered runs of times in the updates. The times *should* be linear (plus sorting) in the inputs, multiplied by the number of times that are not `lt` the time that follows them in sorted order (sounds weird, but `lt` is the partial order, not the total order).

Here are some timings for the first batch using sizes from one to one million.

```
worker 0, round 1 finished after Duration { secs: 0, nanos: 90736 }
worker 0, round 10 finished after Duration { secs: 0, nanos: 215832 }
worker 0, round 100 finished after Duration { secs: 0, nanos: 842466 }
worker 0, round 1000 finished after Duration { secs: 0, nanos: 6758881 }
worker 0, round 10000 finished after Duration { secs: 0, nanos: 55910134 }
worker 0, round 100000 finished after Duration { secs: 0, nanos: 500888218 }
worker 0, round 1000000 finished after Duration { secs: 5, nanos: 405282347 }
```

The average time (per-element) starts at 90us, and drops to just above 5us. It would be great if the improvement were monotonic (the one million number is annoying), but this will take some more detailed measurement.

Happily, this seems to carry over well to multiple (two) workers:

```
worker 0, round 1 finished after Duration { secs: 0, nanos: 88360 }
worker 0, round 10 finished after Duration { secs: 0, nanos: 181533 }
worker 0, round 100 finished after Duration { secs: 0, nanos: 575334 }
worker 0, round 1000 finished after Duration { secs: 0, nanos: 3669810 }
worker 0, round 10000 finished after Duration { secs: 0, nanos: 28990809 }
worker 0, round 100000 finished after Duration { secs: 0, nanos: 284345798 }
worker 0, round 1000000 finished after Duration { secs: 2, nanos: 835915066 }
```

Most seem to have good improvement, and (mysteriously) even the batch size one experiment doesn't suffer.

**Comment by frankmcsherry** [comment]
Fixed by #21.
